{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3478b99f-5162-4b68-9a41-33327ae99794",
   "metadata": {},
   "source": [
    "# Data Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f416acc1-695f-49b9-a649-49a1451c4b0a",
   "metadata": {},
   "source": [
    "## Importing the libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "53484a4e-ec20-496a-8e02-ec08e5c7c7d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1b55f50-f40d-47ae-8413-124c755217be",
   "metadata": {},
   "source": [
    "## Dataset\n",
    "### Showing the Data\n",
    "Show the top columns of the dataset, which includes the data of each cancer tumour."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "fdc00bdb-139d-48df-a645-6b1de9384b74",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\jAlej\\AppData\\Local\\Temp\\ipykernel_23388\\184879827.py:1: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df = pd.read_csv('./data/snv-parse-real1.txt', sep='\\t')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>START_POS_REF</th>\n",
       "      <th>END_POS_REF</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>REF_MFVdVs</th>\n",
       "      <th>ALT_MFVdVs</th>\n",
       "      <th>Sample_Name</th>\n",
       "      <th>FILTER_Mutect2</th>\n",
       "      <th>FILTER_Freebayes</th>\n",
       "      <th>FILTER_Vardict</th>\n",
       "      <th>FILTER_Varscan</th>\n",
       "      <th>m2_MQ</th>\n",
       "      <th>f_MQMR</th>\n",
       "      <th>vs_SSC</th>\n",
       "      <th>vs_SPV</th>\n",
       "      <th>vd_SSF</th>\n",
       "      <th>vd_MSI</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>13110</td>\n",
       "      <td>13110</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G/NA/G/G/</td>\n",
       "      <td>A/NA/A/A/</td>\n",
       "      <td>icgc_cll-T</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>41.91</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.522430</td>\n",
       "      <td>0.23427</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>15015</td>\n",
       "      <td>15015</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G/NA/NA/G/</td>\n",
       "      <td>C/NA/NA/C/</td>\n",
       "      <td>icgc_cll-T</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>43.42</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.302390</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>16949</td>\n",
       "      <td>16949</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>NA/NA/NA/A/</td>\n",
       "      <td>NA/NA/NA/C/</td>\n",
       "      <td>icgc_cll-T</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.023282</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>40552</td>\n",
       "      <td>40552</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>NA/NA/NA/T/</td>\n",
       "      <td>NA/NA/NA/C/</td>\n",
       "      <td>icgc_cll-T</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>26.0</td>\n",
       "      <td>0.002231</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>46907</td>\n",
       "      <td>46907</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>NA/NA/NA/T/</td>\n",
       "      <td>NA/NA/NA/C/</td>\n",
       "      <td>icgc_cll-T</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>17.0</td>\n",
       "      <td>0.017670</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49315</th>\n",
       "      <td>Y</td>\n",
       "      <td>59028157</td>\n",
       "      <td>59028157</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T/NA/NA/NA/</td>\n",
       "      <td>C/NA/NA/NA/</td>\n",
       "      <td>icgc_cll-T</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>54.17</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49316</th>\n",
       "      <td>Y</td>\n",
       "      <td>59028833</td>\n",
       "      <td>59028833</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G/NA/NA/NA/</td>\n",
       "      <td>T/NA/NA/NA/</td>\n",
       "      <td>icgc_cll-T</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>56.34</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49317</th>\n",
       "      <td>Y</td>\n",
       "      <td>59030255</td>\n",
       "      <td>59030255</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G/NA/NA/NA/</td>\n",
       "      <td>A/NA/NA/NA/</td>\n",
       "      <td>icgc_cll-T</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>56.37</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49318</th>\n",
       "      <td>Y</td>\n",
       "      <td>59031439</td>\n",
       "      <td>59031439</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>T/NA/T/T/</td>\n",
       "      <td>A/NA/A/A/</td>\n",
       "      <td>icgc_cll-T</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>54.38</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.525710</td>\n",
       "      <td>0.54983</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49319</th>\n",
       "      <td>Y</td>\n",
       "      <td>59031799</td>\n",
       "      <td>59031799</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T/NA/NA/NA/</td>\n",
       "      <td>C/NA/NA/NA/</td>\n",
       "      <td>icgc_cll-T</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>58.63</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>49320 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Chr  START_POS_REF  END_POS_REF REF ALT   REF_MFVdVs   ALT_MFVdVs  \\\n",
       "0       1          13110        13110   G   A    G/NA/G/G/    A/NA/A/A/   \n",
       "1       1          15015        15015   G   C   G/NA/NA/G/   C/NA/NA/C/   \n",
       "2       1          16949        16949   A   C  NA/NA/NA/A/  NA/NA/NA/C/   \n",
       "3       1          40552        40552   T   C  NA/NA/NA/T/  NA/NA/NA/C/   \n",
       "4       1          46907        46907   T   C  NA/NA/NA/T/  NA/NA/NA/C/   \n",
       "...    ..            ...          ...  ..  ..          ...          ...   \n",
       "49315   Y       59028157     59028157   T   C  T/NA/NA/NA/  C/NA/NA/NA/   \n",
       "49316   Y       59028833     59028833   G   T  G/NA/NA/NA/  T/NA/NA/NA/   \n",
       "49317   Y       59030255     59030255   G   A  G/NA/NA/NA/  A/NA/NA/NA/   \n",
       "49318   Y       59031439     59031439   T   A    T/NA/T/T/    A/NA/A/A/   \n",
       "49319   Y       59031799     59031799   T   C  T/NA/NA/NA/  C/NA/NA/NA/   \n",
       "\n",
       "      Sample_Name  FILTER_Mutect2  FILTER_Freebayes  FILTER_Vardict  \\\n",
       "0      icgc_cll-T            True             False           False   \n",
       "1      icgc_cll-T            True             False           False   \n",
       "2      icgc_cll-T           False             False           False   \n",
       "3      icgc_cll-T           False             False           False   \n",
       "4      icgc_cll-T           False             False           False   \n",
       "...           ...             ...               ...             ...   \n",
       "49315  icgc_cll-T            True             False           False   \n",
       "49316  icgc_cll-T            True             False           False   \n",
       "49317  icgc_cll-T            True             False           False   \n",
       "49318  icgc_cll-T            True             False           False   \n",
       "49319  icgc_cll-T            True             False           False   \n",
       "\n",
       "       FILTER_Varscan  m2_MQ  f_MQMR  vs_SSC    vs_SPV   vd_SSF  vd_MSI  \n",
       "0               False  41.91     NaN     2.0  0.522430  0.23427     2.0  \n",
       "1               False  43.42     NaN     5.0  0.302390      NaN     NaN  \n",
       "2                True    NaN     NaN    16.0  0.023282      NaN     NaN  \n",
       "3                True    NaN     NaN    26.0  0.002231      NaN     NaN  \n",
       "4                True    NaN     NaN    17.0  0.017670      NaN     NaN  \n",
       "...               ...    ...     ...     ...       ...      ...     ...  \n",
       "49315           False  54.17     NaN     NaN       NaN      NaN     NaN  \n",
       "49316           False  56.34     NaN     NaN       NaN      NaN     NaN  \n",
       "49317           False  56.37     NaN     NaN       NaN      NaN     NaN  \n",
       "49318           False  54.38     NaN     2.0  0.525710  0.54983     1.0  \n",
       "49319           False  58.63     NaN     NaN       NaN      NaN     NaN  \n",
       "\n",
       "[49320 rows x 18 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv('./data/snv-parse-real1.txt', sep='\\t')\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b01addf-0b8a-450f-9d92-9c8a6c87d129",
   "metadata": {},
   "source": [
    "Use the method 'info' to show how many Non-null values are they, and what type of data object are they"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "968971bc-f254-4c6f-bd5e-ba05824e467b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 49320 entries, 0 to 49319\n",
      "Data columns (total 18 columns):\n",
      " #   Column            Non-Null Count  Dtype  \n",
      "---  ------            --------------  -----  \n",
      " 0   Chr               49320 non-null  object \n",
      " 1   START_POS_REF     49320 non-null  int64  \n",
      " 2   END_POS_REF       49320 non-null  int64  \n",
      " 3   REF               49320 non-null  object \n",
      " 4   ALT               49320 non-null  object \n",
      " 5   REF_MFVdVs        49320 non-null  object \n",
      " 6   ALT_MFVdVs        49320 non-null  object \n",
      " 7   Sample_Name       49320 non-null  object \n",
      " 8   FILTER_Mutect2    49320 non-null  bool   \n",
      " 9   FILTER_Freebayes  49320 non-null  bool   \n",
      " 10  FILTER_Vardict    49320 non-null  bool   \n",
      " 11  FILTER_Varscan    49320 non-null  bool   \n",
      " 12  m2_MQ             8719 non-null   float64\n",
      " 13  f_MQMR            7537 non-null   float64\n",
      " 14  vs_SSC            39668 non-null  float64\n",
      " 15  vs_SPV            39668 non-null  float64\n",
      " 16  vd_SSF            16076 non-null  float64\n",
      " 17  vd_MSI            16076 non-null  float64\n",
      "dtypes: bool(4), float64(6), int64(2), object(6)\n",
      "memory usage: 5.5+ MB\n"
     ]
    }
   ],
   "source": [
    "df.info()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e13936e9-72db-4909-9034-bf36ca5bbc44",
   "metadata": {},
   "source": [
    "Use the method 'describe' to see the distribution of the numerical values and statistics data. Good to look at min/max values to see any outliers that must put attention on, etc."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c785c812-1259-4913-8b5c-dec3473cbbe4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>START_POS_REF</th>\n",
       "      <th>END_POS_REF</th>\n",
       "      <th>m2_MQ</th>\n",
       "      <th>f_MQMR</th>\n",
       "      <th>vs_SSC</th>\n",
       "      <th>vs_SPV</th>\n",
       "      <th>vd_SSF</th>\n",
       "      <th>vd_MSI</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>4.932000e+04</td>\n",
       "      <td>4.932000e+04</td>\n",
       "      <td>8719.000000</td>\n",
       "      <td>7537.000000</td>\n",
       "      <td>39668.000000</td>\n",
       "      <td>39668.000000</td>\n",
       "      <td>16076.000000</td>\n",
       "      <td>16076.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>6.509624e+07</td>\n",
       "      <td>6.509624e+07</td>\n",
       "      <td>54.510235</td>\n",
       "      <td>49.345667</td>\n",
       "      <td>15.221009</td>\n",
       "      <td>0.118016</td>\n",
       "      <td>0.101967</td>\n",
       "      <td>4.169232</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>5.462033e+07</td>\n",
       "      <td>5.462033e+07</td>\n",
       "      <td>7.070806</td>\n",
       "      <td>15.615680</td>\n",
       "      <td>17.291468</td>\n",
       "      <td>0.169660</td>\n",
       "      <td>0.186070</td>\n",
       "      <td>5.324648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>3.000000e+00</td>\n",
       "      <td>3.000000e+00</td>\n",
       "      <td>10.130000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>2.151707e+07</td>\n",
       "      <td>2.151707e+07</td>\n",
       "      <td>50.680000</td>\n",
       "      <td>43.193500</td>\n",
       "      <td>8.000000</td>\n",
       "      <td>0.018322</td>\n",
       "      <td>0.000110</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>5.135118e+07</td>\n",
       "      <td>5.135118e+07</td>\n",
       "      <td>57.480000</td>\n",
       "      <td>57.583300</td>\n",
       "      <td>12.000000</td>\n",
       "      <td>0.051464</td>\n",
       "      <td>0.024045</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>9.623363e+07</td>\n",
       "      <td>9.623363e+07</td>\n",
       "      <td>60.000000</td>\n",
       "      <td>60.000000</td>\n",
       "      <td>17.000000</td>\n",
       "      <td>0.151600</td>\n",
       "      <td>0.117655</td>\n",
       "      <td>4.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>2.492405e+08</td>\n",
       "      <td>2.492405e+08</td>\n",
       "      <td>62.650000</td>\n",
       "      <td>70.000000</td>\n",
       "      <td>255.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>56.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       START_POS_REF   END_POS_REF        m2_MQ       f_MQMR        vs_SSC  \\\n",
       "count   4.932000e+04  4.932000e+04  8719.000000  7537.000000  39668.000000   \n",
       "mean    6.509624e+07  6.509624e+07    54.510235    49.345667     15.221009   \n",
       "std     5.462033e+07  5.462033e+07     7.070806    15.615680     17.291468   \n",
       "min     3.000000e+00  3.000000e+00    10.130000     0.000000      0.000000   \n",
       "25%     2.151707e+07  2.151707e+07    50.680000    43.193500      8.000000   \n",
       "50%     5.135118e+07  5.135118e+07    57.480000    57.583300     12.000000   \n",
       "75%     9.623363e+07  9.623363e+07    60.000000    60.000000     17.000000   \n",
       "max     2.492405e+08  2.492405e+08    62.650000    70.000000    255.000000   \n",
       "\n",
       "             vs_SPV        vd_SSF        vd_MSI  \n",
       "count  39668.000000  16076.000000  16076.000000  \n",
       "mean       0.118016      0.101967      4.169232  \n",
       "std        0.169660      0.186070      5.324648  \n",
       "min        0.000000      0.000000      0.000000  \n",
       "25%        0.018322      0.000110      1.000000  \n",
       "50%        0.051464      0.024045      2.000000  \n",
       "75%        0.151600      0.117655      4.000000  \n",
       "max        1.000000      1.000000     56.000000  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef0e9e66-f928-4a60-8b83-7f1ad25438ee",
   "metadata": {},
   "source": [
    "The following method allows us to also see the number of null values of each column.\n",
    "\n",
    "We see that m2_MQ and f_MQMR have simimlar missing values (which makes sense beacuse both tell about the mapping quality of the genome), then vs_SSC/SPV have also similar Nan values because both are related to varScan, and finally vd_SSF/MSI have similar NAN values because both are related to varDict. \n",
    "\n",
    "Possible reasons: \n",
    "- If there is high error in a read or the coverage was too low, then maybe outputs were not released.\n",
    "- It could be related to how algorithms work and that in specific cases they decide to not output a given value.\n",
    "- Maybe the researcher deleted those rows when preprocessing?\n",
    "- Maybe if an algorithm didn't stored any SNV position, then skip the data. Maybe they provide metrics at variant side.\n",
    "- Maybe all of the above."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c8ecb74c-5a23-4981-9c52-af6732ae7a0f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Chr                     0\n",
       "START_POS_REF           0\n",
       "END_POS_REF             0\n",
       "REF                     0\n",
       "ALT                     0\n",
       "REF_MFVdVs              0\n",
       "ALT_MFVdVs              0\n",
       "Sample_Name             0\n",
       "FILTER_Mutect2          0\n",
       "FILTER_Freebayes        0\n",
       "FILTER_Vardict          0\n",
       "FILTER_Varscan          0\n",
       "m2_MQ               40601\n",
       "f_MQMR              41783\n",
       "vs_SSC               9652\n",
       "vs_SPV               9652\n",
       "vd_SSF              33244\n",
       "vd_MSI              33244\n",
       "dtype: int64"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.isnull().sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "896774e7-ec4c-490a-b1e9-28448f7bdc39",
   "metadata": {},
   "source": [
    "### Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "501ee79b-5d16-4485-be90-a0a4904d5f83",
   "metadata": {},
   "source": [
    "Print the columns that compose the dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "84bfdf15-1692-4529-8935-86df92a1df0e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Chr',\n",
       " 'START_POS_REF',\n",
       " 'END_POS_REF',\n",
       " 'REF',\n",
       " 'ALT',\n",
       " 'REF_MFVdVs',\n",
       " 'ALT_MFVdVs',\n",
       " 'Sample_Name',\n",
       " 'FILTER_Mutect2',\n",
       " 'FILTER_Freebayes',\n",
       " 'FILTER_Vardict',\n",
       " 'FILTER_Varscan',\n",
       " 'm2_MQ',\n",
       " 'f_MQMR',\n",
       " 'vs_SSC',\n",
       " 'vs_SPV',\n",
       " 'vd_SSF',\n",
       " 'vd_MSI']"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "columns = df.columns.tolist()\n",
    "columns"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01b5515d-879a-416a-8b36-75e4dabc57d1",
   "metadata": {},
   "source": [
    "To be able to see the rows that have null values, we can use the following method"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "85f8482b-be72-4537-9466-c731ec3b1f1e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>START_POS_REF</th>\n",
       "      <th>END_POS_REF</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>REF_MFVdVs</th>\n",
       "      <th>ALT_MFVdVs</th>\n",
       "      <th>Sample_Name</th>\n",
       "      <th>FILTER_Mutect2</th>\n",
       "      <th>FILTER_Freebayes</th>\n",
       "      <th>FILTER_Vardict</th>\n",
       "      <th>FILTER_Varscan</th>\n",
       "      <th>m2_MQ</th>\n",
       "      <th>f_MQMR</th>\n",
       "      <th>vs_SSC</th>\n",
       "      <th>vs_SPV</th>\n",
       "      <th>vd_SSF</th>\n",
       "      <th>vd_MSI</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>10303</td>\n",
       "      <td>10303</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>NA/NA/NA/C/</td>\n",
       "      <td>NA/NA/NA/T/</td>\n",
       "      <td>icgc_cll_tumour</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.277780</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>16949</td>\n",
       "      <td>16949</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A/NA/NA/A/</td>\n",
       "      <td>C/NA/NA/C/</td>\n",
       "      <td>icgc_cll_tumour</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>22.92</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20.0</td>\n",
       "      <td>0.008314</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>20628</td>\n",
       "      <td>20628</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>NA/A/A/A/</td>\n",
       "      <td>NA/G/G/G/</td>\n",
       "      <td>icgc_cll_tumour</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.139060</td>\n",
       "      <td>0.00469</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>30489</td>\n",
       "      <td>30489</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>NA/NA/NA/G/</td>\n",
       "      <td>NA/NA/NA/A/</td>\n",
       "      <td>icgc_cll_tumour</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13.0</td>\n",
       "      <td>0.046735</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>30808</td>\n",
       "      <td>30808</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>NA/NA/NA/A/</td>\n",
       "      <td>NA/NA/NA/G/</td>\n",
       "      <td>icgc_cll_tumour</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>19.0</td>\n",
       "      <td>0.012309</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3535</th>\n",
       "      <td>9</td>\n",
       "      <td>139138662</td>\n",
       "      <td>139138662</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G/NA/NA/NA/</td>\n",
       "      <td>A/NA/NA/NA/</td>\n",
       "      <td>icgc_cll_tumour</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>60.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3536</th>\n",
       "      <td>9</td>\n",
       "      <td>140378948</td>\n",
       "      <td>140378948</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G/NA/NA/NA/</td>\n",
       "      <td>A/NA/NA/NA/</td>\n",
       "      <td>icgc_cll_tumour</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>60.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3537</th>\n",
       "      <td>9</td>\n",
       "      <td>141046294</td>\n",
       "      <td>141046294</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C/NA/NA/NA/</td>\n",
       "      <td>T/NA/NA/NA/</td>\n",
       "      <td>icgc_cll_tumour</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>58.97</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3538</th>\n",
       "      <td>9</td>\n",
       "      <td>141062718</td>\n",
       "      <td>141062718</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T/NA/NA/NA/</td>\n",
       "      <td>C/NA/NA/NA/</td>\n",
       "      <td>icgc_cll_tumour</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>58.84</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3539</th>\n",
       "      <td>9</td>\n",
       "      <td>141081667</td>\n",
       "      <td>141081667</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G/NA/NA/NA/</td>\n",
       "      <td>T/NA/NA/NA/</td>\n",
       "      <td>icgc_cll_tumour</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>46.62</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3470 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Chr  START_POS_REF  END_POS_REF REF ALT   REF_MFVdVs   ALT_MFVdVs  \\\n",
       "0       1          10303        10303   C   T  NA/NA/NA/C/  NA/NA/NA/T/   \n",
       "1       1          16949        16949   A   C   A/NA/NA/A/   C/NA/NA/C/   \n",
       "2       1          20628        20628   A   G    NA/A/A/A/    NA/G/G/G/   \n",
       "3       1          30489        30489   G   A  NA/NA/NA/G/  NA/NA/NA/A/   \n",
       "4       1          30808        30808   A   G  NA/NA/NA/A/  NA/NA/NA/G/   \n",
       "...   ...            ...          ...  ..  ..          ...          ...   \n",
       "3535    9      139138662    139138662   G   A  G/NA/NA/NA/  A/NA/NA/NA/   \n",
       "3536    9      140378948    140378948   G   A  G/NA/NA/NA/  A/NA/NA/NA/   \n",
       "3537    9      141046294    141046294   C   T  C/NA/NA/NA/  T/NA/NA/NA/   \n",
       "3538    9      141062718    141062718   T   C  T/NA/NA/NA/  C/NA/NA/NA/   \n",
       "3539    9      141081667    141081667   G   T  G/NA/NA/NA/  T/NA/NA/NA/   \n",
       "\n",
       "          Sample_Name  FILTER_Mutect2  FILTER_Freebayes  FILTER_Vardict  \\\n",
       "0     icgc_cll_tumour           False             False           False   \n",
       "1     icgc_cll_tumour           False             False           False   \n",
       "2     icgc_cll_tumour           False             False            True   \n",
       "3     icgc_cll_tumour           False             False           False   \n",
       "4     icgc_cll_tumour           False             False           False   \n",
       "...               ...             ...               ...             ...   \n",
       "3535  icgc_cll_tumour            True             False           False   \n",
       "3536  icgc_cll_tumour            True             False           False   \n",
       "3537  icgc_cll_tumour            True             False           False   \n",
       "3538  icgc_cll_tumour            True             False           False   \n",
       "3539  icgc_cll_tumour            True             False           False   \n",
       "\n",
       "      FILTER_Varscan  m2_MQ  f_MQMR  vs_SSC    vs_SPV   vd_SSF  vd_MSI  \n",
       "0               True    NaN     NaN     5.0  0.277780      NaN     NaN  \n",
       "1               True  22.92     NaN    20.0  0.008314      NaN     NaN  \n",
       "2              False    NaN     NaN     8.0  0.139060  0.00469     2.0  \n",
       "3               True    NaN     NaN    13.0  0.046735      NaN     NaN  \n",
       "4               True    NaN     NaN    19.0  0.012309      NaN     NaN  \n",
       "...              ...    ...     ...     ...       ...      ...     ...  \n",
       "3535           False  60.00     NaN     NaN       NaN      NaN     NaN  \n",
       "3536           False  60.00     NaN     NaN       NaN      NaN     NaN  \n",
       "3537           False  58.97     NaN     NaN       NaN      NaN     NaN  \n",
       "3538           False  58.84     NaN     NaN       NaN      NaN     NaN  \n",
       "3539           False  46.62     NaN     NaN       NaN      NaN     NaN  \n",
       "\n",
       "[3470 rows x 18 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df.isnull().any(axis=1)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d73d5af4-46b7-4e3e-a3f2-41a60c0346ca",
   "metadata": {},
   "source": [
    "## Understanding the data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "44382f22-537d-4927-9010-ea466080f311",
   "metadata": {},
   "source": [
    "The following code will be used to understand what each columns mean and what do their value represent"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "81c125c8-0b38-4845-b7e8-868c2cc655e4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Chr', 'START_POS_REF', 'END_POS_REF', 'REF', 'ALT', 'REF_MFVdVs', 'ALT_MFVdVs', 'Sample_Name', 'FILTER_Mutect2', 'FILTER_Freebayes', 'FILTER_Vardict', 'FILTER_Varscan', 'm2_MQ', 'f_MQMR', 'vs_SSC', 'vs_SPV', 'vd_SSF', 'vd_MSI']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>START_POS_REF</th>\n",
       "      <th>END_POS_REF</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>REF_MFVdVs</th>\n",
       "      <th>ALT_MFVdVs</th>\n",
       "      <th>Sample_Name</th>\n",
       "      <th>FILTER_Mutect2</th>\n",
       "      <th>FILTER_Freebayes</th>\n",
       "      <th>FILTER_Vardict</th>\n",
       "      <th>FILTER_Varscan</th>\n",
       "      <th>m2_MQ</th>\n",
       "      <th>f_MQMR</th>\n",
       "      <th>vs_SSC</th>\n",
       "      <th>vs_SPV</th>\n",
       "      <th>vd_SSF</th>\n",
       "      <th>vd_MSI</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>13110</td>\n",
       "      <td>13110</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G/NA/G/G/</td>\n",
       "      <td>A/NA/A/A/</td>\n",
       "      <td>icgc_cll-T</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>41.91</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.522430</td>\n",
       "      <td>0.23427</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>15015</td>\n",
       "      <td>15015</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G/NA/NA/G/</td>\n",
       "      <td>C/NA/NA/C/</td>\n",
       "      <td>icgc_cll-T</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>43.42</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.302390</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>16949</td>\n",
       "      <td>16949</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>NA/NA/NA/A/</td>\n",
       "      <td>NA/NA/NA/C/</td>\n",
       "      <td>icgc_cll-T</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.023282</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>40552</td>\n",
       "      <td>40552</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>NA/NA/NA/T/</td>\n",
       "      <td>NA/NA/NA/C/</td>\n",
       "      <td>icgc_cll-T</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>26.0</td>\n",
       "      <td>0.002231</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>46907</td>\n",
       "      <td>46907</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>NA/NA/NA/T/</td>\n",
       "      <td>NA/NA/NA/C/</td>\n",
       "      <td>icgc_cll-T</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>17.0</td>\n",
       "      <td>0.017670</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chr  START_POS_REF  END_POS_REF REF ALT   REF_MFVdVs   ALT_MFVdVs  \\\n",
       "0   1          13110        13110   G   A    G/NA/G/G/    A/NA/A/A/   \n",
       "1   1          15015        15015   G   C   G/NA/NA/G/   C/NA/NA/C/   \n",
       "2   1          16949        16949   A   C  NA/NA/NA/A/  NA/NA/NA/C/   \n",
       "3   1          40552        40552   T   C  NA/NA/NA/T/  NA/NA/NA/C/   \n",
       "4   1          46907        46907   T   C  NA/NA/NA/T/  NA/NA/NA/C/   \n",
       "\n",
       "  Sample_Name  FILTER_Mutect2  FILTER_Freebayes  FILTER_Vardict  \\\n",
       "0  icgc_cll-T            True             False           False   \n",
       "1  icgc_cll-T            True             False           False   \n",
       "2  icgc_cll-T           False             False           False   \n",
       "3  icgc_cll-T           False             False           False   \n",
       "4  icgc_cll-T           False             False           False   \n",
       "\n",
       "   FILTER_Varscan  m2_MQ  f_MQMR  vs_SSC    vs_SPV   vd_SSF  vd_MSI  \n",
       "0           False  41.91     NaN     2.0  0.522430  0.23427     2.0  \n",
       "1           False  43.42     NaN     5.0  0.302390      NaN     NaN  \n",
       "2            True    NaN     NaN    16.0  0.023282      NaN     NaN  \n",
       "3            True    NaN     NaN    26.0  0.002231      NaN     NaN  \n",
       "4            True    NaN     NaN    17.0  0.017670      NaN     NaN  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "columns = df.columns.tolist()\n",
    "print(columns)\n",
    "df[df.isnull().any(axis=1)]\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "627f58e2-fabd-4656-8bbd-83e4256ce8e2",
   "metadata": {},
   "source": [
    "## Understanding the columns\n",
    "\n",
    "Breakdown of each column\n",
    "\n",
    "---\n",
    "\n",
    "## **1. Genomic Position Information**\n",
    "| Column | Description | Example Value |\n",
    "|--------|------------|--------------|\n",
    "| **Chr** | Chromosome number where the SNV is located. | `1` |\n",
    "| **START_POS_REF** | Start position of the SNV in the reference genome. | `13110` |\n",
    "| **END_POS_REF** | End position of the SNV in the reference genome (same as `START_POS_REF` for SNVs because an SNV is a single nucleotide that was modified). | `13110` |\n",
    "\n",
    "- **Format:** Integer (`1, 2, ..., 22, X, Y` for chromosomes)\n",
    "- **Purpose:** Identifies where in the genome this SNV occurs.\n",
    "\n",
    "---\n",
    "\n",
    "## **2. Variant Information**\n",
    "| Column | Description | Example Value |\n",
    "|--------|------------|--------------|\n",
    "| **REF** | Reference nucleotide (what is expected in a normal genome). | `G` |\n",
    "| **ALT** | Alternate nucleotide (mutation found in the tumor sample). | `A` |\n",
    "\n",
    "- **Format:** Single-character DNA base (`A, T, C, G`)\n",
    "- **Purpose:** Shows how the sequence has changed in cancer cells.\n",
    "\n",
    "---\n",
    "\n",
    "## **3. Allele Information Across Algorithms**\n",
    "| Column | Description | Example Value |\n",
    "|--------|------------|--------------|\n",
    "| **REF_MFVdVs** | Reference allele calls from all four algorithms: MuTect2 (M), FreeBayes (F), VarDict (Vd), and VarScan (Vs). | `G/NA/G/G/` |\n",
    "| **ALT_MFVdVs** | Alternate allele calls from all four algorithms: MuTect2 (M), FreeBayes (F), VarDict (Vd), and VarScan (Vs). | `A/NA/A/A/` |\n",
    "\n",
    "In the example above, it is telling us that MuTect2, VarDict, and VarScan expected the reference nucleotie (what is expected in normal genome) to be 'G', but the alterate allele was 'A', meaning it is a SNV. On the other hand, FreeBayes didn't disagreed that a mutation exists.\n",
    "\n",
    "- **Format:** Slash-separated (`M/F/Vd/Vs/`), `NA` means the algorithm didn't detect a variant at this position.\n",
    "- **Purpose:** Shows how each SNV caller interpreted the mutation.\n",
    "\n",
    "---\n",
    "\n",
    "## **4. Sample Information**\n",
    "| Column | Description | Example Value |\n",
    "|--------|------------|--------------|\n",
    "| **Sample_Name** | The tumor sample ID. | `icgc_cll-T` |\n",
    "\n",
    "- **Format:** String (sample name)\n",
    "- **Purpose:** Identifies which tumor sample this variant belongs to.\n",
    "\n",
    "---\n",
    "\n",
    "## **5. Variant Detection by Each Algorithm**\n",
    "| Column | Description | Example Value |\n",
    "|--------|------------|--------------|\n",
    "| **FILTER_Mutect2** | `TRUE` if MuTect2 passed the variant as significant, `FALSE` otherwise. | `TRUE` |\n",
    "| **FILTER_Freebayes** | `TRUE` if FreeBayes passed the variant, `FALSE` otherwise. | `FALSE` |\n",
    "| **FILTER_Vardict** | `TRUE` if VarDict passed the variant, `FALSE` otherwise. | `FALSE` |\n",
    "| **FILTER_Varscan** | `TRUE` if VarScan passed the variant, `FALSE` otherwise. | `FALSE` |\n",
    "\n",
    "- **Format:** Boolean (`TRUE` or `FALSE`)\n",
    "- **Purpose:** Indicates whether each algorithm considers the SNV a real mutation.\n",
    "\n",
    "---\n",
    "\n",
    "## **6. Variant-Specific Features (Algorithm-Specific Quality Scores)**\n",
    "| Column | Description | Example Value |\n",
    "|--------|------------|--------------|\n",
    "| **m2_MQ** | Mapping quality from **MuTect2** (how confidently the read aligns to the genome). | `41.91` |\n",
    "| **f_MQMR** | Mapping quality from **FreeBayes**. | `NA` |\n",
    "| **vs_SSC** | Strand bias from **VarScan** (how evenly reads support the mutation in both DNA strands). | `2` |\n",
    "| **vs_SPV** | Statistical p-value from **VarScan** (significance of the mutation). | `0.52243` |\n",
    "| **vd_SSF** | Somatic status score from **VarDict**. | `0.23427` |\n",
    "| **vd_MSI** | Microsatellite instability (MSI) from **VarDict**. | `2` |\n",
    "\n",
    "- **Format:** Float or `NA` (if the algorithm didn't calculate it).\n",
    "- **Purpose:** Algorithm-specific quality metrics for filtering out false positives.\n",
    "\n",
    "---\n",
    "\n",
    "## **Summary**\n",
    "- **Columns 1-3** → Where the variant is in the genome.\n",
    "- **Columns 4-5** → What the mutation is.\n",
    "- **Columns 6-7** → How each algorithm sees the reference and alternate alleles.\n",
    "- **Column 8** → Which tumor sample the variant belongs to.\n",
    "- **Columns 9-12** → Whether each algorithm validated this SNV.\n",
    "- **Columns 13-18** → Algorithm-specific scores for quality assessment."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ebdfb863-f2cc-48c8-b5d4-d525c1ed306b",
   "metadata": {},
   "source": [
    "## Results\n",
    "### Showing and Understanding the data\n",
    "Show the top columns of the dataset in the '.bed' represent the true somatic SNVs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "4a71d363-1fda-44f6-8a23-3c5ed5a0f34e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>2180985</td>\n",
       "      <td>2180985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>5035185</td>\n",
       "      <td>5035185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>8881322</td>\n",
       "      <td>8881322</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>8929624</td>\n",
       "      <td>8929624</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>9196716</td>\n",
       "      <td>9196716</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1314</th>\n",
       "      <td>X</td>\n",
       "      <td>142275497</td>\n",
       "      <td>142275497</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1315</th>\n",
       "      <td>X</td>\n",
       "      <td>144737796</td>\n",
       "      <td>144737796</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1316</th>\n",
       "      <td>Y</td>\n",
       "      <td>6407994</td>\n",
       "      <td>6407994</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1317</th>\n",
       "      <td>Y</td>\n",
       "      <td>15154839</td>\n",
       "      <td>15154839</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1318</th>\n",
       "      <td>Y</td>\n",
       "      <td>15681833</td>\n",
       "      <td>15681833</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1319 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      0          1          2\n",
       "0     1    2180985    2180985\n",
       "1     1    5035185    5035185\n",
       "2     1    8881322    8881322\n",
       "3     1    8929624    8929624\n",
       "4     1    9196716    9196716\n",
       "...  ..        ...        ...\n",
       "1314  X  142275497  142275497\n",
       "1315  X  144737796  144737796\n",
       "1316  Y    6407994    6407994\n",
       "1317  Y   15154839   15154839\n",
       "1318  Y   15681833   15681833\n",
       "\n",
       "[1319 rows x 3 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv('./real1/real1_truth.bed', sep='\\t', header=None)\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c4ae2b2a-9ca3-4f83-abfe-2be835ee402d",
   "metadata": {},
   "source": [
    "Each BED file contains **three columns**, and each row in a BED file represents a **confirmed somatic SNV**.\n",
    "\n",
    "| Column  | Description                                | Example  |\n",
    "|---------|--------------------------------------------|---------|\n",
    "| **Chr** | Chromosome number                         | `1`     |\n",
    "| **Start** | Start position of the SNV               | `13110` |\n",
    "| **End** | End position of the SNV *(same as Start for SNVs)* | `13110` |"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3095fa93-3dc1-4619-8027-91bc0922b333",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
